src/device_trezor/trezor/protocol.cpp

```diff
@@ -0,0 +1,896 @@
+    if (in_memory()){
+      return;
+    }
+  }
+
+  std::shared_ptr<messages::monero::MoneroTransactionAllInputsSetRequest> Signer::step_all_inputs_set(){
+    return std::make_shared<messages::monero::MoneroTransactionAllInputsSetRequest>();
+  }
+
+  void Signer::step_all_inputs_set_ack(std::shared_ptr<const messages::monero::MoneroTransactionAllInputsSetAck> ack){
+    if (is_offloading()){
+      // If offloading, expect rsig configuration.
+      if (!ack->has_rsig_data()){
+        throw exc::ProtocolException("Rsig offloading requires rsig param");
+      }
+
+      auto & rsig_data = ack->rsig_data();
+      if (!rsig_data.has_mask()){
+        throw exc::ProtocolException("Gamma masks not present in offloaded version");
+      }
+
+      auto & mask = rsig_data.mask();
+      if (mask.size() != 32 * num_outputs()){
+        throw exc::ProtocolException("Invalid number of gamma masks");
+      }
+
+      m_ct.rsig_gamma.reserve(num_outputs());
+      for(size_t c=0; c < num_outputs(); ++c){
+        rct::key cmask{};
+        memcpy(cmask.bytes, mask.data() + c * 32, 32);
+        m_ct.rsig_gamma.emplace_back(cmask);
+      }
+    }
+  }
+
+  std::shared_ptr<messages::monero::MoneroTransactionSetOutputRequest> Signer::step_set_output(size_t idx){
+    m_ct.cur_output_idx = idx;
+    m_ct.cur_output_in_batch_idx += 1;   // assumes sequential call to step_set_output()
+
+    auto res = std::make_shared<messages::monero::MoneroTransactionSetOutputRequest>();
+
+    CHECK_AND_ASSERT_THROW_MES(idx < m_ct.tx_data.splitted_dsts.size(), "Invalid transaction index");
+    CHECK_AND_ASSERT_THROW_MES(idx < m_ct.tx_out_entr_hmacs.size(), "Invalid transaction index");
+    auto & cur_dst = m_ct.tx_data.splitted_dsts[idx];
+    translate_dst_entry(res->mutable_dst_entr(), &cur_dst);
+    res->set_dst_entr_hmac(m_ct.tx_out_entr_hmacs[idx]);
+
+    // Range sig offloading to the host
+    if (!is_offloading()) {
+      return res;
+    }
+
+    CHECK_AND_ASSERT_THROW_MES(m_ct.cur_batch_idx < m_ct.grouping_vct.size(), "Invalid batch index");
+    if (m_ct.grouping_vct[m_ct.cur_batch_idx] > m_ct.cur_output_in_batch_idx) {
+      return res;
+    }
+
+    auto rsig_data = res->mutable_rsig_data();
+    auto batch_size = m_ct.grouping_vct[m_ct.cur_batch_idx];
+
+    if (!is_req_bulletproof()){
+      if (batch_size > 1){
+        throw std::invalid_argument("Borromean cannot batch outputs");
+      }
+
+      CHECK_AND_ASSERT_THROW_MES(idx < m_ct.rsig_gamma.size(), "Invalid gamma index");
+      rct::key C{}, mask = m_ct.rsig_gamma[idx];
+      auto genRsig = rct::proveRange(C, mask, cur_dst.amount);  // TODO: rsig with given mask
+      auto serRsig = cn_serialize(genRsig);
+      m_ct.tx_out_rsigs.emplace_back(genRsig);
+      rsig_data->set_rsig(serRsig);
+
+    } else {
+      std::vector<uint64_t> amounts;
+      rct::keyV masks;
+      for(size_t i = 0; i < batch_size; ++i){
+        const size_t bidx = 1 + idx - batch_size + i;
```

Comment:
idx is a parameter, no check for >= anything, so if it's, say, 0, bidx = 1+0-batch_size+0. Seems like it could underflow at first glance ? Maybe that's intentional since it gets caught by the < below since unsigned underflow is OK ?

Reply:
I've added the explicit underflow check to be sure.